omnigan/trainer.py

```diff
@@ -1179,10 +1195,39 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
                             self.domain_labels[batch_domain],
                             self.D["m"]["Advent"],
                         )
-
-                        disc_loss["m"]["Advent"] += (
-                            self.opts.train.lambdas.advent.adv_main * loss_main
-                        )
+                        if self.opts.dis.m.gan_type == "GAN" or "WGAN_norm":
+                            disc_loss["m"]["Advent"] += (
+                                self.opts.train.lambdas.advent.adv_main * loss_main
+                            )
+                        elif self.opts.dis.m.gan_type == "WGAN":
+                            for p in self.D["m"]["Advent"].parameters():
+                                p.data.clamp_(
+                                    self.opts.dis.m.wgan_clamp_lower,
+                                    self.opts.dis.m.wgan_clamp_upper,
+                                )
+                            disc_loss["m"]["Advent"] += (
+                                self.opts.train.lambdas.advent.adv_main * loss_main
+                            )
+                        elif self.opts.dis.m.gan_type == "WGAN_gp":
+                            prob_need_grad = autograd.Variable(prob, requires_grad=True)
+                            d_out = self.D["m"]["Advent"](prob_need_grad)
+                            # github code reference: https://github.com/caogang/wgan-gp/blob/master/gan_cifar10.py
+                            grads = autograd.grad(
+                                outputs=d_out,
```

Comment:
Can you move that to a `tutils` function to keep this loop as clean as possible?